homeassistant/components/synology_dsm/backup.py

```diff
@@ -0,0 +1,219 @@
+"""Support for Synology DSM backup agents."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Callable, Coroutine
+import logging
+from typing import Any
+
+from synology_dsm.api.file_station import SynoFileStation
+from synology_dsm.exceptions import SynologyDSMAPIErrorException
+
+from homeassistant.components.backup import AgentBackup, BackupAgent, BackupAgentError
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.aiohttp_client import ChunkAsyncStreamIterator
+from homeassistant.helpers.json import json_dumps
+from homeassistant.util.json import JsonObjectType, json_loads_object
+
+from .const import (
+    CONF_BACKUP_PATH,
+    CONF_BACKUP_SHARE,
+    DATA_BACKUP_AGENT_LISTENERS,
+    DOMAIN,
+)
+from .models import SynologyDSMData
+
+LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    if not (
+        entries := hass.config_entries.async_loaded_entries(DOMAIN)
+    ) or not hass.data.get(DOMAIN):
+        LOGGER.debug("No proper config entry found")
+        return []
+    agents: list[BackupAgent] = []
+    for entry in entries:
+        syno_data: SynologyDSMData = hass.data[DOMAIN][entry.unique_id]
+        if syno_data.api.file_station and entry.options.get(CONF_BACKUP_PATH):
+            agents.append(SynologyDSMBackupAgent(hass, entry))
+    return agents
+
+
+@callback
+def async_register_backup_agents_listener(
+    hass: HomeAssistant,
+    *,
+    listener: Callable[[], None],
+    **kwargs: Any,
+) -> Callable[[], None]:
+    """Register a listener to be called when agents are added or removed.
+
+    :return: A function to unregister the listener.
+    """
+    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)
+
+    @callback
+    def remove_listener() -> None:
+        """Remove the listener."""
+        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
+        if not hass.data[DATA_BACKUP_AGENT_LISTENERS]:
+            hass.data.pop(DATA_BACKUP_AGENT_LISTENERS)
```

Comment:
```suggestion
            del hass.data[DATA_BACKUP_AGENT_LISTENERS]
```

nit: del is a bit faster than `pop` since it avoids the function call and we don't need a default here